std/algorithm/searching.d

```diff
@@ -1204,6 +1210,75 @@ bool endsWith(alias pred, R)(R doesThisEnd)
     }
 }
 
+/**
+Iterates the passed range and selects the extreme element with $(D less).
+If the extreme element occurs multiple time, the first occurrence will be
+returned.
+
+Params:
+    map = custom accessor for the comparison key
+    selector = custom mapping for the extrema selection
+    r = Range from which the extreme value will be selected
+
+Returns: The extreme value according to $(D less) of the passed-in values.
+*/
+private auto extremum(alias map = "a", alias selector = "a < b", Range)(Range r)
+    if (isInputRange!Range && !isInfinite!Range)
+in
+{
+    assert(!r.empty, "r is an empty range");
+}
+body
+{
+    alias mapFun = unaryFun!map;
+    alias selectorFun = binaryFun!selector;
+
+    // init extremum
+    Unqual!(ElementType!Range) extremeElement = r.front;
+    alias MapType = Unqual!(typeof(mapFun((ElementType!Range).init)));
+    MapType extremeElementMapped = mapFun(r.front);
+
+    static if (isRandomAccessRange!Range && hasLength!Range)
+    {
+        immutable size_t lenArray = r.length;
+        for (size_t i = 1; i < lenArray; i++)
```

Comment:
No, actually I think that merging the input and random-access range paths has no disadvantages:
```
alias Element = ElementType!Range;
alias MapType = Unqual!(typeof(mapFun(Element.init)));

// init extremum
Unqual!Element extremeElement = r.front;
MapType extremeElementMapped = mapFun(extremeElement);

r.popFront();
foreach (const element; r)
{
    MapType elementMapped = mapFun(element);
    if (selectorFun(elementMapped, extremeElementMapped))
    {
        extremeElement = element;
        extremeElementMapped = elementMapped;
    }
}

return extremeElement;
```